### Import Modules and Data

In [211]:
# Import pandas and numpy
import pandas as pd
import numpy as np

In [212]:
file = 'data/si-salaries.csv'
df = pd.read_csv(file)

### Data cleaning

In [213]:
# Preview data
df.head()

Employee Name                 Job Title   Base Pay   Overtime Pay  \
0      James Hammond            Superintendent  355285.00            0.0   
1   Craig M Guensler            Superintendent  233539.86            0.0   
2        Michael Lin            Superintendent  398979.02            0.0   
3  Joel Lee Kinnamon  Superintendent/President  399006.00            0.0   
4    Mary E Streshly            Superintendent  192784.38            0.0   

    Other Pay   Benefits   Total Pay & Benefits    Year  Notes  \
0   197112.00  167573.00              719970.00  2020.0   None   
1    10099.42  342424.16              586063.44  2020.0   None   
2    57050.69   80208.48              536238.19  2020.0   None   
3    60457.00   66557.00              526020.00  2020.0   None   
4   259544.88   66560.32              518889.58  2020.0   None   

                  Agency  Status  
0      Ontario-Montclair     NaN  
1   Wheatland Union High     NaN  
2   Corona-Norco Unified     NaN  
3  College of the Desert     NaN  
4     Sequoia Union High     NaN

In [214]:
# Check column names
df.columns

Index(['Employee Name', ' Job Title', ' Base Pay', ' Overtime Pay',
       ' Other Pay', ' Benefits', ' Total Pay & Benefits', ' Year', ' Notes',
       ' Agency', ' Status'],
      dtype='object')

In [215]:
# Rename the columns; strip the whitespace
for i in df.columns:
    df.rename(columns={i:i.strip()}, inplace=True)

# Verify change
df.columns

Index(['Employee Name', 'Job Title', 'Base Pay', 'Overtime Pay', 'Other Pay',
       'Benefits', 'Total Pay & Benefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [216]:
# Look at counts of 'Job Title'
df['Job Title'].value_counts()

Superintendent                                          435
Assistant Superintendent                                 68
Deputy Superintendent                                    52
Associate Superintendent                                 23
Superintendent/President                                 21
                                                       ... 
Assistant Superintendent Student Support Services         1
Assistant Superintendent-Bsvc                             1
Assistant Superintendent, Pupil Services &amp; Selpa      1
Assistant Superintendent-Edser                            1
Assistant Superintendent Of Instructional Services        1
Name: Job Title, Length: 255, dtype: int64

In [217]:
# Find school related superintedent jobs only
df['Job Title'].unique()

array(['Superintendent', 'Superintendent/President',
       'Superintendent Of Schools', 'District Superintendent',
       'Superintendent President', 'President/Superintendent',
       'County Superintendent Of Schl', 'Superintendent-President',
       'Deputy Superintendent, Business Services And Operations',
       'General Superintendent', 'Associate Superintendent',
       'County Superintendent/Schools', 'Assistant Superintendent Moft',
       'Deputy Superintendent - 2302', 'Street Tree Superintendent',
       'State Park Superintendent Iii', 'Associate Superintendent - Cbo',
       'State Park Superintendent V', 'Shops Superintendent',
       'Deputy Superintendent Of Business Services',
       'Assistant Superintendent',
       'President/Superintendent Chief Executive Officer',
       'Water Utility Superintendent', 'Deputy Superintendent',
       'A Superintendent', 'Superintendent, Deputy',
       'Superintendent/Presi', 'Superintendent-Cert',
       'Interim Superintendent

NOTE: Only looking for superintendents that work in education

In [218]:
# Create a list with education related superintendents
titles = [
        'Superintendent', 
        'Superintendent/President',
        'Superintendent Of Schools', 
        'District Superintendent',
        'Superintendent President', 
        'President/Superintendent',
        'County Superintendent Of Schl', 
        'Superintendent-President', 
        'Superintendent-District'
        ]

# Only get rows that are titles
df = df[df['Job Title'].isin(titles)].reset_index().drop(columns=['index'])

# Drop 'Job Title'
df.drop(columns= 'Job Title', axis=1, inplace=True)

# Check shape of data
df.shape

(490, 10)

In [219]:
# Check the count of 'Notes'
df.value_counts('Notes')

Notes
None    490
dtype: int64

Since there is no useful information in the notes column, drop the column.

In [220]:
# Drop notes column
df.drop(columns= 'Notes', axis=1, inplace=True)

In [221]:
# Check 'Year' column
df['Year'].value_counts()

2020.0    490
Name: Year, dtype: int64

Since the data contains information only for 2020, delete the 'Year' column

In [222]:
# Drop 'Year'
df.drop(columns= 'Year', axis=1, inplace=True)

In [223]:
# Check 'Status'
df['Status']

0      NaN
1      NaN
2      NaN
3      NaN
4      NaN
      ... 
485    NaN
486    NaN
487    NaN
488    NaN
489    NaN
Name: Status, Length: 490, dtype: object

In [224]:
# Count missing values in 'Status'
df['Status'].isna().sum()

490

Since all rows are missing in the 'Status' column, drop it

In [225]:
# Drop 'Status'
df.drop(columns= 'Status', axis=1, inplace=True)

Check for any duplicate names in 'Employee Name'

In [226]:
# Print counts of 'Employee Name'
df['Employee Name'].value_counts()

Not Provided            3
Craig M Guensler        2
Michael R Watenpaugh    2
Krystal K Lomanto       2
Adela Jones             1
                       ..
Kimberly Mae Fricker    1
Stacy L McAfee          1
Juan Cruz               1
Kimberly M Berman       1
Shannon M Sanford       1
Name: Employee Name, Length: 485, dtype: int64

In [227]:
# Locate rows with "Craig M Guensler"
df[df['Employee Name'] == 'Craig M Guensler']

Employee Name   Base Pay  Overtime Pay  Other Pay   Benefits  \
1    Craig M Guensler  233539.86           0.0   10099.42  342424.16   
257  Craig M Guensler  233539.86           0.0   10099.42   49392.44   

     Total Pay & Benefits                     Agency  
1               586063.44       Wheatland Union High  
257             293031.72  Wheatland School District

I decide to keep the first row because 'Total Pay & Benefits' is higher.

In [228]:
# Drop duplicate row
df = df.drop(df.index[257])
df.reset_index(inplace=True, drop=True)

In [229]:
# Locate rows with "Not Provided"
df[df['Employee Name'] == 'Not Provided']

Employee Name   Base Pay  Overtime Pay  Other Pay  Benefits  \
48   Not Provided  310000.00           0.0   18500.00  70866.00   
222  Not Provided  134238.15           0.0  150228.55  20443.45   
264  Not Provided  227718.66           0.0    6000.00  57150.11   

     Total Pay & Benefits            Agency  
48              399366.00  Glendale Unified  
222             304910.15   Hayward Unified  
264             290868.77   Lincoln Unified

I decide to leave all rows with no 'Employee Name' because the pay and agency are not similar.

In [230]:
# Locate rows with "Michael R Watenpaugh"
df[df['Employee Name'] == 'Michael R Watenpaugh']

Employee Name   Base Pay  Overtime Pay  Other Pay  Benefits  \
270  Michael R Watenpaugh  226763.27           0.0    5624.63  56979.68   
382  Michael R Watenpaugh  172581.20           0.0   34222.00  45477.13   

     Total Pay & Benefits                       Agency  
270             289367.58  Cotati-Rohnert Park Unified  
382             252280.33         San Rafael City High

I decide to drop the row with the lower 'Total Pay & Benefits'.

In [231]:
# Drop duplicate row
df = df.drop(df.index[382])
df.reset_index(inplace=True, drop=True)

In [232]:
# Locate rows with 'Krystal K Lomanto'
df[df['Employee Name'] == 'Krystal K Lomanto']

Employee Name   Base Pay  Overtime Pay  Other Pay  Benefits  \
476  Krystal K Lomanto  179345.16           0.0     2160.0  45228.24   
477  Krystal K Lomanto  179345.16           0.0     2160.0  45228.24   

     Total Pay & Benefits                                 Agency  
476              226733.4                        San Benito High  
477              226733.4  San Benito County Office of Education

The information is the same, so I decide to keep the first row.

In [233]:
# Drop duplicate row
df = df.drop(df.index[477])
df.reset_index(inplace=True, drop=True)

In [234]:
# Verify
df['Employee Name'].value_counts()

Not Provided                              3
James Hammond                             1
Edward President\&#39;S Office Knudson    1
Christi Lynn Barrett                      1
Thomas Rooney                             1
                                         ..
Kimberly Mae Fricker                      1
Stacy L McAfee                            1
Juan Cruz                                 1
Kimberly M Berman                         1
Shannon M Sanford                         1
Name: Employee Name, Length: 485, dtype: int64

In [235]:
df.shape

(487, 7)

No need for 'Employee Name' anymore. Therefore, drop the column

In [236]:
# Drop 'Employee Name'
df.drop(columns='Employee Name', inplace=True)

In [237]:
# Count values for 'Overtime Pay'
df['Overtime Pay'].value_counts()

0.00        479
500.00        1
3834.78       1
36536.46      1
20642.00      1
7411.00       1
19933.60      1
2500.00       1
9231.28       1
Name: Overtime Pay, dtype: int64

In [238]:
# Check data types
df.dtypes

Base Pay                float64
Overtime Pay            float64
Other Pay               float64
Benefits                float64
Total Pay & Benefits    float64
Agency                   object
dtype: object

In [239]:
# Convert 'Agency' into string
df['Agency'] = df.Agency.astype('string')

In [240]:
# Create new column 'Total Pay'
df['Total Pay'] = (
    df['Base Pay'] + 
    df['Overtime Pay'] +
    df['Other Pay']
)

In [241]:
# Convert float64 columns into int
for i in df.columns:
    if df[i].dtype == 'float64':
        df[i] = df[i].astype('int')
        
# Verify
df.dtypes

Base Pay                 int64
Overtime Pay             int64
Other Pay                int64
Benefits                 int64
Total Pay & Benefits     int64
Agency                  string
Total Pay                int64
dtype: object

In [242]:
# Rearrange column order
column_order = [
                'Agency', 
                'Base Pay',
                'Other Pay', 
                'Overtime Pay', 
                'Total Pay', 
                'Benefits',
                'Total Pay & Benefits'
                ]

df = df.loc[:, column_order]

In [243]:
# Verify change
df.head()

Agency  Base Pay  Other Pay  Overtime Pay  Total Pay  \
0      Ontario-Montclair    355285     197112             0     552397   
1   Wheatland Union High    233539      10099             0     243639   
2   Corona-Norco Unified    398979      57050             0     456029   
3  College of the Desert    399006      60457             0     459463   
4     Sequoia Union High    192784     259544             0     452329   

   Benefits  Total Pay & Benefits  
0    167573                719970  
1    342424                586063  
2     80208                536238  
3     66557                526020  
4     66560                518889

### Feature Engineering

In [244]:
# Import Caliofornia_Counties
data = 'data/ca-cities.csv'
cities = pd.read_csv(data)

In [245]:
# Preview data
cities.head(3)

Name   Latitude   Longitude
0      Adelanto  34.582769 -117.409214
1  Agoura Hills  34.153339 -118.761675
2       Alameda  37.765206 -122.241636

In [246]:
# Check for missing values
cities.isna().any()

Name         False
Latitude     False
Longitude    False
dtype: bool

In [247]:
# Check data types
cities.dtypes

Name          object
Latitude     float64
Longitude    float64
dtype: object

In [248]:
# Convert city names to string
cities['Name'] = cities['Name'].astype('string')

# Confirm change
cities.dtypes

Name          string
Latitude     float64
Longitude    float64
dtype: object

In [249]:
# Import process from fuzzywuzzy; string similarity
from fuzzywuzzy import process

# Convert city names to a Pandas Series
choices = pd.Series(cities.Name)

si_cities = []
latitudes = []
longitudes = []

# String similarity
for i, row in df.iterrows():
    string = df.loc[i, 'Agency']
    choice = process.extract(string, choices, limit=1)
    
    city = choice[0][0]
    
    si_cities.append(city)

# Create new column called 
df['City'] = si_cities

In [250]:
# Check quality of matching
df[['Agency', 'City']]

Agency             City
0                    Ontario-Montclair        Montclair
1                 Wheatland Union High        Wheatland
2                 Corona-Norco Unified           Corona
3                College of the Desert      Palm Desert
4                   Sequoia Union High       Union City
..                                 ...              ...
482                      Banta Unified     Santa Monica
483            Rescue Union Elementary       Union City
484  Alexander Valley Union Elementary  Fountain Valley
485                      Coastline ROP         Coalinga
486        Stanislaus Union Elementary       Union City

[487 rows x 2 columns]

In [251]:
# Save to CSV
df.to_csv('data/uncleaned-si-cities.csv')

### Note: 
Some Agency rows got matched poorly, so I had to verify and clean the data manually in the csv file — ended up saving it to a file called <b> cleaned-matched-cities</b> and went from <b> 488 rows</b> to <b> 334 rows </b> (~75% accuracy).

In [252]:
# Import data 
data = 'data/cleaned-si-cities.csv'
cleaned_si_cities = pd.read_csv(data)

In [253]:
# Merge cities and cleaned_si_cities
final_si_data = cleaned_cities.merge(cities, left_on='City', right_on='Name')

# Drop 'Name' column
final_si_data.drop('Name', axis=1, inplace=True)

# Verify change
final_si_data.head()

Agency  Base Pay  Other Pay  Overtime Pay  Total Pay  \
0      Ontario-Montclair    355285     197112             0     552397   
1   Wheatland Union High    233539      10099             0     243639   
2   Corona-Norco Unified    398979      57050             0     456029   
3   Corona-Norco Unified    252334      39041             0     291375   
4  College of the Desert    399006      60457             0     459463   

   Benefits  Total Pay & Benefits         City   Latitude   Longitude  
0    167573                719970    Montclair  34.077511 -117.689778  
1    342424                586063    Wheatland  39.009894 -121.423014  
2     80208                536238       Corona  33.875294 -117.566439  
3     52337                343713       Corona  33.875294 -117.566439  
4     66557                526020  Palm Desert  33.722244 -116.374456

In [255]:
# Save to CSV
final_si_data.to_csv('data/final_si_data.csv')

### To see visualization, head over to: 
https://public.tableau.com/app/profile/jordanlarot/viz/SuperintendentTotalPayBenefitsinCalifornia/TotalPayBenefits